### ISSUE TITLE
Minimum length for fragments of hyphenated words

### ISSUE THREAD

**Issue by jchtt** [koreader/koreader#3848]
* KOReader version: v2015.11-1564-gbfd55c7_2018-03-13
* Device: Kobo H2O

#### Issue

Would it be possible to include settings for the minimal length of splits when hyphenating words, for example in the way it is done in the [Hyphenate This](https://www.mobileread.com/forums/showthread.php?t=208534) plugin for calibre? There, you can enter something like
```
LEFTHYPHENMIN = 5
RIGHTHYPHENMIN = 5
```
which leads to words not being broken if either the left or the right part would be shorter than 5 characters. With that setting, it would be admissible to split

> extra-ordinary,

but not

> ex-traordinary,

nor

> extraordi-nary.

Regardless of this, thanks a lot for developing koreader!

**Comment by poire-z** [comment]
MAX_REAL_WORD seems not used, and WORD_LENGTH just sets some char table length.
But src/hyphman.cpp is quite small, and there's the word "Tex" (that should appeal to you :) in many places.

There's this that could looks like it's for that:
https://github.com/koreader/crengine/blob/a09aac18672f5f9c57d87457afbac205001ed793/crengine/src/hyphman.cpp#L652-L654
that was used there
https://github.com/koreader/crengine/blob/a09aac18672f5f9c57d87457afbac205001ed793/crengine/src/hyphman.cpp#L728-L739

But we'd rather first get into how all that works... There's different code for 3 different hyph dict.
May be there's stuff already defined in some of these different hyph dict specs to solve this problem? (which would mean we could hack the dict files instead of the crengine code?)

**Comment by Frenzie** [comment]
Theoretically one could attack the basic problem in the dictionary, but in practice you would set `\lefthyphenmin=2` and `\righthyphenmin=3`. (Or rather I think Babel does. Those are the defaults for English.) But LaTeX always chooses bigger chunks unless it doesn't see a better choice.

Hyphenation in (La)TeX is very nice, I think there's nothing that justifies as well as (La)TeX, except other things that use the same algorithm. (It has not, to my knowledge, been improved upon since the '80s.). But you still fine-tune it when you're done.

```tex
% okay, in real life that's more like the problem you're trying to avoid :-P
\hyphenation{h-yphenate}
```

In my thesis, one wrongful breakup slipped in due to a last minute edit somewhere in the vicinity. Grmbl.

PS In English one might want to use the UK hyphenation patterns. They're slightly less "let's break up the word!" than the US ones.

**Comment by jchtt** [comment]
Thanks for looking into this! The problem was most apparent to me in German, where some common verb endings look a bit ugly when you brake them down to two final characters.

I’m not quite sure what those pdb files contain. The pattern files at least are not word based, but just contain patterns (hence the name I guess), so I cannot just easily change the breakpoints of words as needed but would need to change something in the algorithm.

**Comment by Frenzie** [comment]
Seems to be one of the old ones: https://github.com/koreader/crengine/tree/925bae617252aa036ea875700dbec46661796f47/cr3gui/data/hyph

Also se https://github.com/koreader/crengine/pull/78 and https://github.com/koreader/crengine/pull/80#issuecomment-325446830

I've never converted one of these patterns, but I'd probably recommend it with an updated one from https://www.tug.org/tex-hyphen/

**Comment by cramoisi** [comment]
@poire-z  @Frenzie  yup ! the value is something the css code (epub3) or the hyph_dic code should do, not something to put in the engine's code because these values change should change by languages or by user's tastes. Theses values are in the hyp_lang.dic on kobo side but not in the pattern files used now by koreader.

I noticed some months ago they are present in the Apache FOP project when it comes to translate the pattern hyph dic to a xml pattern https://xmlgraphics.apache.org/fop/2.1/hyphenation.html

**Comment by Frenzie** [comment]
"except other things that use the same algorithm"

> Apache™ FOP uses Liang's hyphenation algorithm, well known from TeX.

Neat! ;-)

**Comment by cramoisi** [comment]
@poire-z : I got two hyphen patterns for french. The one koreader provide, which you convert from TeX with the program provide by crengine (if I'm remember correctly and if it was you) and mine, which I converted manually from lo-oo-ressources-linguistiques-fr-v5.7.zip package (https://extensions.libreoffice.org/extensions/dictionnaires-francais/5.7) using the UTF-8 hyph_fr.dic file.

Considering only the number of pages of an epub when applying koreader's or mine (Ok, i need a better test-case it's for that I speak of it here), I can tell mine is performing better (I gain 5 pages on a regular 300 pages epub without changing anything, so I could infer that) but it's hard to pinpoint _where_ it performs better.

As far as I can tell, the difference in the files are  : 

1. I keep the dot markers (which doesn't seem to harm anything) - but koreader get rid of them for a space and that, I don't understand.
2. I keep all the second part of the file with the non straight single quote ‘ which is heavily used in french for apostrophe. This part is not present at all in the TeX file.
3. I keep all the third part with dash

In fact, by manual conversion, I just mean I add the xml tag on each line.
I still got some word to fine tune (like "aujourd'hui") but theses doesn't hyphenate with any of the patterns I applied.

[hyph_fr.dic.zip](https://github.com/koreader/koreader/files/1897388/hyph_fr.dic.zip)

**Comment by poire-z** [comment]
Yes, I did update the french hyph dict in https://github.com/koreader/crengine/pull/78.
So, if I were to believe what I wrote there, it came from LibreOffice (dunno how and why it differs from yours.)

I don't know anything at all about these files syntax and how that works, so thanks for the little overview.
I won't dig much into that word splitting science (cause I'm not that interested), so I'll trust your conclusions.
Can you provide your final `French.pattern` so I can just see how it does? I won't probably _notice_ any change cause I'm not that sensitive to word splitting, but that would be a _GO_ to replace mine with yours if I don't :)

Is there such good/updated hyph dicts in latest Libre Office from all languages? Cause may be it would be good to keep them up-to-date with a live source (Libre Office)?
Could you try finding/making a german one for @jchtt to test their quality?

**Comment by cramoisi** [comment]
@poire-z Will attach them to this comment latter this morning. I'm still asking myself about the question of the dot. In TeX they are the mark of the beginning and the ending of a word. When you convert the hyph TeX file they were replaced with a space. I noticed no differences in the result and, if it's really the case, I have the tendency to prefer the dot because they are explicits. Is there any clues about this in the code ? (german, french, dutch, at least the main difference is about the straight and non straight sigle quote. The actual french file "shouldn't" do well with words when a non-straight quote comes before. And it's more or less a duplicate of all lines concerned by this. I also noticed the fop pattern is corrupted : all the quotes are non-straight. Another question : could it be better, like in the fop pattern, to have them sorted for better readability ? even though i don't see another way to do it other than by hands)

**Comment by poire-z** [comment]
Dots are indeed explicitely replaced by space by the converter in https://github.com/koreader/crengine/blob/a09aac18672f5f9c57d87457afbac205001ed793/crengine/Tools/HyphConv/mkpattern.cpp#L55-L61

I guess it's because, that way, it's easier to match word boundaries that usually are spaces (althought I don't see why it couldn't be dots): https://github.com/koreader/crengine/blob/a09aac18672f5f9c57d87457afbac205001ed793/crengine/src/hyphman.cpp#L681-L684

About the various single quotes, I don't see anything related to them in the quote, nor in the hyphenation patterns doc on the Apache FOP page, so I guess they match litteral quotes?
In that case, if there's just the angled single quote `’` (does it have the same equivalence as `'` in all languages? are there other such single quote clones?), we could replace them with the regular single quote just in the previous hyphman quoted code above.

> could it be better, like in the fop pattern, to have them sorted for better readability ? 

I guess it's better to not change them to see changes vs upstream new releases and review/apply changes. But if you choose to become the Master of the Hyph Dicts :) , you become the upstream and can do whatever you want :)

**Comment by cramoisi** [comment]
@poire-z  The problem I see here is at least for french, a lot of boundaries are not "simple" space (Unicode speaking, u+0020). There is a lot of non breakable spaces with ponctuation. like u+00a0 or u+202F. Do you think it would conflict with the "space boundary" in use now ?

**Comment by poire-z** [comment]
> Do you think it would conflict with the "space boundary" in use now ? In the case we couldn't be sure, I would opt for the dot.

I don't think it matters in the way it works:
The converter replace dot with spaces.
crengine hyph code gets each word (from some other part of the code that I think accurately splits a line into words, splitting on multiple different spaces and punctuation - if there is a problem there, it's to be fixed elsewhere and is not related to hyphenation). For each single word, it prepends and appends a space, and simply look if each pattern literally appears into this modified word. So, the patterns with dot-replaced-by-space work as expected.

I get it matters to you when visually looking at the final file to see dots instead of spaces, but if we were to change the prepend and append of spaces to dots, that would need updating all the other hyph patterns.

**Comment by cramoisi** [comment]
For french : it should be better for apostrophe and for composed words.

[French_with_dots.zip](https://github.com/koreader/koreader/files/1901997/French_with_dots.zip)
[French_without dots.zip](https://github.com/koreader/koreader/files/1902003/French_without.dots.zip)

**Comment by poire-z** [comment]
If I get these things right:
- the dot is something specific in the TeX hyph dict specs to indicate it should match a word boundary
- the only job of HyphConv is actually to replace these dots with spaces (just because crengine itselfs add spaces to start and end of the words it is asked to hyphenate) - it's a technical indifferent choice, it could have in both decided to use any ridiculous character (like `£`) it is sure to not have in the words it is given.
- some other part of crengine code splits a line into words, and I can tell you (see below) it is this part of the code that splits `aujourd'hui` or `aujourd’hui` or `aujourd-hui` into `aujourd` and `hui` - and asks the hyphenation code to hyphenate `aujourd` - and the hyphenation code decided for `au[SPLIT]jourd` on my sample text).
- the hyphenation code just do string comparisons, so, if it adds space to start and end of words, it needs to have a space in the pattern for the word boundary match to work.

So, if what I said above is correct, your `French_with_dots.zip` would work exactly as if all the patterns with dots were removed (because the hyph code won't get any word with dots, so none of your patterns with dots would ever match anything)

For information, here's some extracts from crengine debug output:

While parsing a hyph dict:
```
INFO Selecting hyphenation dictionary ./data/hyph/French.pattern
DEBUG Pattern: (2'2) ''' - 22
DEBUG Pattern: ( a4) ' a' - 004
DEBUG Pattern: ('a4) ''a' - 004
DEBUG Pattern: ( â4) ' â' - 004
DEBUG Pattern: ('â4) ''â' - 004
DEBUG Pattern: (ab2h) 'abh' - 0020
DEBUG Pattern: ( ab3réa) ' abréa' - 0003000
DEBUG Pattern: ('ab3réa) ''abréa' - 0003000
DEBUG Pattern: (ad2h) 'adh' - 0020
DEBUG Pattern: (a1è2dre) 'aèdre' - 012000
DEBUG Pattern: ( ae3s4ch) ' aesch' - 0003400
DEBUG Pattern: ('ae3s4ch) ''aesch' - 0003400
DEBUG Pattern: (1alcool) 'alcool' - 1000000
DEBUG Pattern: (a2l1algi) 'alalgi' - 0210000
DEBUG Pattern: ( amino1a2c) ' aminoac' - 000000120
DEBUG Pattern: ('amino1a2c) ''aminoac' - 000000120
DEBUG Pattern: ( ana3s4tr) ' anastr' - 00003400
DEBUG Pattern: ('ana3s4tr) ''anastr' - 00003400
DEBUG Pattern: (1a2nesthési) 'anesthési¿¿0' - 120000000
DEBUG Pattern: ( anti1a2) ' antia' - 0000012
DEBUG Pattern: ('anti1a2) ''antia' - 0000012
```

While rendering some pages, and the words it is asked to hyphenate:
```
DEBUG Render is finished

TRACE word to hyphenate: ' reconnus '
DEBUG Pattern matched: re 100 on reconnus  0000000000
DEBUG Pattern matched: co 100 on connus  00000000
DEBUG Pattern matched: nu 100 on nus  00000
TRACE Hyphenate: re-con-nu|s  r0e1-c0o0n1-n0u0|s0

TRACE word to hyphenate: ' influence '
DEBUG Pattern matched:  i 004 on  influence  000000000000
DEBUG Pattern matched: fl 120 on fluence  000000000
DEBUG Pattern matched: lu 100 on luence  20000000
DEBUG Pattern matched: ce  4000 on ce  0000
DEBUG Pattern matched: ce  4000 on ce  4000
DEBUG Pattern matched: ce 100 on ce  4000
TRACE Hyphenate: in-fluenc|e  i4n1-f2l0u0e0n4c0|e0

TRACE word to hyphenate: ' femmes '
DEBUG Pattern matched: fe 100 on femmes  00000000
DEBUG Pattern matched: mes  40000 on mes  00000
DEBUG Pattern matched: me 100 on mes  40000
TRACE Hyphenate: femme|s  f0e0m4m0e0|s0

TRACE word to hyphenate: ' aujourd '
DEBUG Pattern matched:  a 004 on  aujourd  0000000000
DEBUG Pattern matched: j 10 on jourd  0000000
TRACE Hyphenate: au-jourd  a4u1-j0o0u0r0d0

TRACE word to hyphenate: ' secrète '
DEBUG Pattern matched: se 100 on secrète  000000000
DEBUG Pattern matched: cr 120 on crète  0000000
DEBUG Pattern matched: rè 100 on rète  200000
DEBUG Pattern matched: te  4000 on te  0000
DEBUG Pattern matched: te  4000 on te  4000
DEBUG Pattern matched: te 100 on te  4000
TRACE Hyphenate: se-crète  s0e1-c2r0è4t0e0

TRACE word to hyphenate: ' invention '
DEBUG Pattern matched:  i 004 on  invention  000000000000
DEBUG Pattern matched: ve 100 on vention  000000000
DEBUG Pattern matched: ti 100 on tion  000000
TRACE Hyphenate: in-ven-|tion  i4n1-v0e0n1-|t0i0o0n0

TRACE word to hyphenate: ' auteurs '
DEBUG Pattern matched:  a 004 on  auteurs  0000000000
DEBUG Pattern matched: te 100 on teurs  0000000
TRACE Hyphenate: au-teurs|  a4u1-t0e0u0r0s0|

TRACE word to hyphenate: ' personnalité '
DEBUG Pattern matched:  per 00040 on  personnalité  000000000000000
DEBUG Pattern matched: pe 100 on personnalité  00400000000000
DEBUG Pattern matched: so 100 on sonnalité  00000000000
DEBUG Pattern matched: na 100 on nalité  00000000
DEBUG Pattern matched: li 100 on lité  000000
DEBUG Pattern matched: té 100 on té  0000
TRACE Hyphenate: per-son-na-|li-té  p0e4r1-s0o0n1-n0a1-|l0i1-t0é0
```

I'm not eager at all to understand what all that means :) but it seems to indicate that your way with dots may be a wrong track.

Note that if in the original text, I put a `l'` before `auteurs` and it reads `l'auteurs`, the above logs stays with `TRACE word to hyphenate: ' auteurs '`. Same if I use this https://www.fileformat.info/info/unicode/char/2019/index.htm instead. Same if I put a `-` instead.
So, you won't solve your apostrophe or composed word problems on the hyphenation side, neither in hyphman.cpp code nor in your pattern files. Any pattern with a single quote would simply never match anything as it is never fed a word with an apostrophe.

The thing is that crengine does not use the whole Tex Pattern hyphenation rules. It just grabs the string patterns, remove the numbers, and do some simple string comparison.
If you put quotes or minus sign in your patterns, these patterns will simply never match anything.

> this line is missing in koreader's actuel converted hyph french file.

And It does not even care about it :) HyphConv indeed does not bother to add it, and crengine, while parsing this file, do some simple string parsing: it does not parse it with some XML engine, so it does not bother it that it is not valid XML.
Guess this was just quickly made to have some _enough-ly well_ working hyphenation in place.

**Comment by cramoisi** [comment]
@Frenzie thanks for all the explanation.
So all the `'` lines in the hyph file could be removed ? So I guess the differences I got from the https://www.dicollecte.org/ file and koreader's should come from additional standard patterns. I will strip the files and make a diff.

> And It does not even care about it :)
I noticed that :-)

**Comment by Frenzie** [comment]
@cramoisi I think you typod there. :-P

**Comment by cramoisi** [comment]
as usual :)

**Comment by jchtt** [comment]
@cramoisi For starters, it should be the `hyph_de_DE/hyph_de_DE.dic` file in the oxt package [here](https://extensions.libreoffice.org/extensions/german-de-de-frami-dictionaries/2017-01-12).

Thanks :)!

**Comment by cramoisi** [comment]
@jchtt  here it is : 
[hyph_de_DE.zip](https://github.com/koreader/koreader/files/1903234/hyph_de_DE.zip)

change [this line](https://github.com/koreader/crengine/blob/3a74635bb8ea412f202929e06d231ed4bb015f51/cr3gui/data/hyph/languages.json#L74) with the filename and drop it in the ../data/hyph/ folder :)

**Comment by cramoisi** [comment]
@Frenzie you're not the same person ? ;-P

**Comment by poire-z** [comment]
(You don't need me to give you logs with ur with_dot.pattern anymore, is that right?)

> It's curious because if i remove all the spaces and the duplicate line, koreader doesn't like it

Dunno why it wouldn't like it. What spaces did you remove? (in the final pattern files, there should be spaces where dots were)

>  the final pattern with nothing not use by koreader is hyph_fr.zip 

It works for me (I don't really see any difference indeed). What do you mean with _with nothing not use by koreader_  ?
(I'll have to test it on my kobo and read a few pages to confirm it don't change much)

Btw, for your  "aujourd'hui" pattern: now that we know that crengine gives to the hyphenation code the _word_ `aujourd` to hyphenate, you may manually add one pattern specifically for that so it's split where you want it: `au-jour-d`

PS: it's often hard to follow @cramoisi mass-multi-edited posts :), and I don't see anything on GitHub to show diff between posts' edits. Is there something that I missed?

**Comment by poire-z** [comment]
> Can you get an hypen on the word "comment" ? It should be " com-ment ", it's at ln 252.

Now, I can't.
Some bits of the related logs:

```
2018/04/12 17:29:55.9367 DEBUG Pattern: (co2g3ni1ti) 'cogniti' - 00230100
2018/04/12 17:29:55.9367 DEBUG Pattern: ( com1me) ' comme' - 0000100
2018/04/12 17:29:55.9367 DEBUG Pattern: ( com3ment ) ' comment ¿¿0' - 000030000
2018/04/12 17:29:55.9367 DEBUG Pattern: (com1pé) 'compé' - 000100
2018/04/12 17:29:55.9367 DEBUG Pattern: (compé1te) 'compéte' - 00000100
2018/04/12 17:29:55.9367 DEBUG Pattern: (compé3tent ) 'compétent¿¿0' - 000003000
2018/04/12 17:29:55.9368 DEBUG Pattern: ( con4) ' con' - 00004
2018/04/12 17:29:55.9368 DEBUG Pattern: (con1fi) 'confi' - 000100
2018/04/12 17:29:55.9368 DEBUG Pattern: (confi1de) 'confide' - 00000100
2018/04/12 17:29:55.9368 DEBUG Pattern: (confi3dent ) 'confident¿¿0' - 000003000
2018/04/12 17:29:55.9368 DEBUG Pattern: (con1ni) 'conni' - 000100
2018/04/12 17:29:55.9368 DEBUG Pattern: (conni1ve) 'connive' - 00000100
2018/04/12 17:29:55.9368 DEBUG Pattern: (conni3vent ) 'connivent¿¿0' - 000003000
2018/04/12 17:29:55.9368 DEBUG Pattern: ( cons4) ' cons' - 000004
2018/04/12 17:29:55.9368 DEBUG Pattern: (con1ti) 'conti' - 000100
2018/04/12 17:29:55.9369 DEBUG Pattern: (conti1ne) 'contine' - 00000100
2018/04/12 17:29:55.9369 DEBUG Pattern: (conti3nent ) 'continent¿¿0' - 000003000

2018/04/12 17:29:55.9880 TRACE word to hyphenate: ' comment '
2018/04/12 17:29:55.9880 DEBUG Pattern matched:  comme 0000100 on  comment  0000000000
2018/04/12 17:29:55.9880 DEBUG Pattern matched:  co 0100 on  comment  0000100000
2018/04/12 17:29:55.9881 DEBUG Pattern matched:  c 004 on  comment  0100100000
2018/04/12 17:29:55.9881 DEBUG Pattern matched: co 100 on comment  140100000
2018/04/12 17:29:55.9881 DEBUG Pattern matched: omment  00200000 on omment  40100000
2018/04/12 17:29:55.9881 DEBUG Pattern matched: omme 00100 on omment  40200000
2018/04/12 17:29:55.9881 DEBUG Pattern matched: me 100 on ment  200000
2018/04/12 17:29:55.9881 TRACE Hyphenate: comment|  c4o0m2m0e0n0t0|

```
Looks like there may be a bug with words ending with spaces, as we see them finally ending with this strange `¿¿0`, and so obviously, they won't match with that garbage added...

**Comment by cramoisi** [comment]
@Frenzie The rubbish is in the test book or in the file ?

**Comment by poire-z** [comment]
@Frenzie and @poire-z are really not the same person, I promise, and I explicitely chosed a light gray logo so that confusion with @Frenzie 's darker one should never happen! :)

It's in hyphman.cpp code (I'm having a look at it), and it happens with my old French.pattern too. And it does not happen on all patterns ending with spaces:
In a latin1 terminal:
```
2018/04/12 17:40:44.8214 DEBUG Pattern: (om2ment ) 'omment ' - 00200000
2018/04/12 17:40:44.8214 DEBUG Pattern: ( com3ment ) ' comment ü°~L~C~@°ü°~L~C~@³0' - 000030000
2018/04/12 17:40:44.8214 DEBUG Pattern: (Ã´2ment ) 'Ã´ment ' - 0200000
2018/04/12 17:40:44.8214 DEBUG Pattern: (slalo2ment ) 'slalomentü°~L~C~@°ü°~L~C~H°0' - 000002000
2018/04/12 17:40:44.8214 DEBUG Pattern: (chro2ment ) 'chroment ü°~L~C~@°ü°~L~C~@²0' - 000020000
2018/04/12 17:40:44.8214 DEBUG Pattern: (to2ment ) 'toment ' - 00200000
```
These are bytes `fc b08c 8380 b0fc b08c 8380 b330`...

**Comment by Frenzie** [comment]
Probably, but I would take a peek at the TeX/LibreOffice patterns for Turkish and Finnish before saying that with any certainty. :-)

**Comment by cramoisi** [comment]
@Frenzie : https://ctan.org/tex-archive/language/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/  or http://ftp.math.purdue.edu/mirrors/ctan.org/language/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/


they are amazingly shorts... but anyway, the longest word could be use as a pattern... so If it's not a problem it should be interesting to keep a long value https://en.wikipedia.org/wiki/Longest_words ; (like 100 ?)

**Comment by cramoisi** [comment]
Tweedehandsemotorverkoopsmannevakbondstakingsvergaderingsameroeperstoespraakskrywerspersverklaringuitreikingsmediakonferensieaankondiging is my new favorite word !

**Comment by poire-z** [comment]
Note that this MAX_PATTERN_SIZE is used to reserve that memory size amount for each pattern (even it the pattern is shorter):
```cpp
class TexPattern {
public:
    lChar16 word[MAX_PATTERN_SIZE];
    char attr[MAX_PATTERN_SIZE+1];
```
So, with a Dutch.pattern containing 12000 patterns, `MAX_PATTERN_SIZE=100` would made for 12000 * 100 * (2+1) = 3,6 MB of RAM usage instead of 360KB.
Which is OK, but wasted. Also, these patterns are also made to be parts of words, so these big big words are probably matched and splitted by smaller patterns, and may be there's no real occurence of big words in any Tex pattern...

**Comment by cramoisi** [comment]
yep, it's the pattern principle to not use a full word as a pattern anyway. In french the longest is 16 and it's a full word with special hyphen rules (" chè2vre3feuille ") so... somewhere between 20 and 40 ?

**Comment by poire-z** [comment]
```
for f in *pattern* ; do echo $f ; grep '<pattern>' $f | awk '{ print length, $0 }' | sort -n | cut -d" " -f2- | tail -1 ; echo ;  done

Catalan.pattern
  <pattern>ru1t2herford</pattern>

Dutch.pattern
  <pattern>zwets5te </pattern>

French.pattern
<pattern> chÃ¨2vre3feuil1le</pattern>

French.pattern.CRAMOISI
<pattern> chÃ¨2vre3feuil1le</pattern>

French.pattern.ORIG
  <pattern> chÃ¨2vre3feuille</pattern>

Galician.pattern
  <pattern>supe3r2a3bilidade</pattern>

Greek.pattern
  <pattern>á½~AÏ~@Î¿Î¹Î¿Ï~E2Ï²1Ï~DÎ¹Î½Î±2Ï²1Î¿á¿¦Î½ </pattern>

Norwegian.pattern
  <pattern>dgjÃ¸6re7n6h</pattern>

Polish.pattern
        <pattern>g8d8z8i8e9n8i8e9g8d8z8i8e1</pattern>

Spanish.pattern
  <pattern> de3s4in3ter2e3sÃ³</pattern>

Turkish.pattern
  <pattern>m1t4rak</pattern>

```
Which makes Greek having the longest pattern of 40 bytes, but only 20 unicode chars.
And Polish having the longest pattern of 26 unicode chars.
Dunno about all the other pdb files...

**Comment by cramoisi** [comment]
pdb file are bound to become pattern files anyway, let's go with 30

**Comment by Frenzie** [comment]
> yep, it's the pattern principle to not use a full word as a pattern anyway.

Hm, but the correct way to split a made-up word like "Tweedehandsemotorverkoopsmannevakbondstakingsvergaderingsameroeperstoespraakskrywerspersverklaringuitreikingsmediakonferensieaankondiging" is on the individual word parts.

Good:

Tweedehandsemotor-verkoopsmannevakbondstakingsvergaderingsameroeperstoespraakskrywerspersverklaringuitreikingsmediakonferensieaankondiging

Bad:

Tweedehandsemo-torverkoopsmannevakbondstakingsvergaderingsameroeperstoespraakskrywerspersverklaringuitreikingsmediakonferensieaankondiging

(Admittedly they probably don't worry too much about made up words.)

PS I'd say the "popular" (not to mention better) form of "Hottentottensoldatententententoonstellingsbouwterrein" is merely "hottentottententententoonstelling." It's not just shorter, but I think throwing soldiers in the middle and a construction ground at the end murders the alliterative effect.

**Comment by Frenzie** [comment]
Just double it to something like 18 or 20 for now? Plus a comment explaining what that value is all about.

**Comment by poire-z** [comment]
Please hold, I'm trying to make it ignore and tell us about oversized words, so we can easily know when that happen.

**Comment by cramoisi** [comment]
gosh, I press the button when reading your message

**Comment by poire-z** [comment]
MAX_PATTERN_SIZE actually means MAX_PATTERN_WORD_SIZE.
There is no limit about the size of the `<pattern>st3uff4 </pattern>` in the XML file.
But when it is parsed, only the letters (and the spaces) account for the `word`.

For that 26 unicode chars polish pattern `<pattern>g8d8z8i8e9n8i8e9g8d8z8i8e1</pattern>`, there are 13 numbers, and only 13 unicode chars that will make the word `gdzieniegdzie` stored in `word` which has its size limited to MAX_PATTERN_SIZE (that's why it would make more sense if it were named MAX_PATTERN_WORD_SIZE).
```
class TexPattern {
public:
    lChar16 word[MAX_PATTERN_SIZE];
    char attr[MAX_PATTERN_SIZE+1];
```
I actually have to limit it to `MAX_PATTERN_SIZE-1` for the last char to be `0` and marks string termination so we don't get fetching some other chars beyond the end of the lChar16 array (that was the garbage we saw).

I want crengine to tell us when crengine logging is enabled:
```
2018/04/12 21:07:59.8842 DEBUG Pattern: (to1yo2) 'toyo' - 00102
2018/04/12 21:07:59.8842 DEBUG Pattern: (turboo2d3rzut) 'turboodr' - 000000230
2018/04/12 21:07:59.8842 WARN Pattern overflowed (10 > 8) and ignored: (turboo2d3rzut) 'turboodr'
2018/04/12 21:07:59.8842 DEBUG Pattern: (tu1) 'tu' - 001
2018/04/12 21:07:59.8843 DEBUG Pattern: (tu2r1b) 'turb' - 00210
2018/04/12 21:24:35.3346 DEBUG Pattern: ( nieprze3d4ramat) ' nieprze' - 000000003
2018/04/12 21:24:35.3347 WARN Pattern overflowed (13 > 8) and ignored: ( nieprze3d4ramat) ' nieprze'
2018/04/12 21:24:35.3762 DEBUG Pattern: (g8d8z8i8e9n8i8e9g8d8z8i8e1) 'gdzienie' - 088889889
2018/04/12 21:24:35.3762 WARN Pattern overflowed (12 > 8) and ignored: (g8d8z8i8e9n8i8e9g8d8z8i8e1) 'gdzienie'
```
and in normal reader mode, including on devices, just:
`CRE WARNING: some hyphenation patterns were too long and have been ignored: increase MAX_PATTERN_SIZE from 9 to 14`
so we know what we need to do.

(I still have to get my counting right)

**Comment by cramoisi** [comment]
Ok !  But when you got that what will you do to get the value ?

**Comment by poire-z** [comment]
It's not for people (or may be for others like you playing with custom hyph dicts), it's for us: we'll need to select every hyph dict and look for this warning, so we know what value to use. It's for us tomorrow and a given for the future hyph dicts update we'll do.

I don't think I'll get to have something I'm 100% sure of this evening :( Can you wait till saturday's nightly?

(We also need to bump cache format or FORMATTING_VERSION_ID in src/lvtinydom.cpp so rendering of cached book is redone - I witnessed some text written over some other text when just changing hyph dicts with the same name on a book with cache this afternoon).

**Comment by cramoisi** [comment]
Great ! (and of course I can wait ... and on top of that, I'm grateful for the good work ;) I'm also kinda happy my whereabouts with the hyph all this day drove out a bug. 

(playing [minit](https://www.gog.com/game/minit) and reading  [Kraken](https://en.wikipedia.org/wiki/Kraken_(novel)) while people are working)

**Comment by poire-z** [comment]
OK, done.
@cramoisi I let you update your PR with the optimal value and increment FORMATTING_VERSION_ID
Please test it with your pattern works (i think you got the emulator working), just to confirm it's still ok.

(I had a ridiculous hard time being sure about arrays bounds, like is `word[MAX_PATTERN_SIZE]` the last usable slot or already overflowed... - too much used to lua where arrays starts at 1...)

**Comment by cramoisi** [comment]
@poire-z I don't have my main computer this morning, but what about using this time to convert some pattern where there is still pdb files ?

**Comment by poire-z** [comment]
Feel free to. I guess .pattern from recent LibreOffice can only be better than old .pdb. Dunno about the Russian_EnGB & Russian_EnUS.

**Comment by cramoisi** [comment]
@poire-z what about taking them from [ctan.org]( https://ctan.org/tex-archive/language/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/ ?)

**Comment by poire-z** [comment]
I know nothing about the TeX ecosystem, so I let @Frenzie answers :)

**Comment by cramoisi** [comment]
The files from libre/open office are all converted from the tex files, in fact ; (the tex files are even packaged in the extension if you look at them) so the source seems to me more reliable and accessible than downloading , unpacking, converting from apache website...

**Comment by cramoisi** [comment]
@poire-z shouldn't I wait for a merge of your modifications for testing ?

**Comment by cramoisi** [comment]
here are bg, cs, da, en-gb, en-us, fi

[hyph.zip](https://github.com/koreader/koreader/files/1906300/hyph.zip)

**Comment by poire-z** [comment]
> @poire-z shouldn't I wait for a merge of your modifications for testing ?

You can apply https://github.com/koreader/crengine/pull/146.diff to your current crengine while it's not yet merged (then, once it's merged: `git reset --hard ; git pull`)

**Comment by poire-z** [comment]
~~git reset --hard ; git pull~~  risks erasing your hard work on hyph patterns that are in cr3gui/data/hyph/, so don't do that :)
Use instead: `git checkout crengine/src/hyphman.cpp ; git pull`

**Comment by cramoisi** [comment]
@poire-z it seems I'm too lame to do use this beyond the very basic (and I say "very" because I know it's basic) ; so if you don't have the time for a step by step I will wait a merge :)

**Comment by poire-z** [comment]
^ That's because it's not yet bumped into `base` (and `base` not yet into `koreader`).
We'll wait for your crengine PR (MAX_PATTERN_SIZE  9 > 16 + FORMATTING_VERSION_ID) to bump them both.

**Comment by Frenzie** [comment]
We should really do something about that. Something like this mostly does the trick:

diff --git a/Makefile.third b/Makefile.third
index d8401a9..26bd8fc 100644
--- a/Makefile.third
+++ b/Makefile.third
@@ -107,7 +107,7 @@ $(DJVULIBRE_LIB): $(JPEG_LIB) $(THIRDPARTY_DIR)/djvulibre/CMakeLists.txt
 
```diff
 # crengine, fetched via GIT as a submodule
 $(CRENGINE_LIB): $(ZLIB) $(PNG_LIB) $(FREETYPE_LIB) $(JPEG_LIB) $(NANOSVG_HEADERS) \
-               $(THIRDPARTY_DIR)/kpvcrlib/crengine $(THIRDPARTY_DIR)/kpvcrlib/CMakeLists.txt
+               $(THIRDPARTY_DIR)/kpvcrlib/crengine/crengine/src $(THIRDPARTY_DIR)/kpvcrlib/CMakeLists.txt
        install -d $(CRENGINE_BUILD_DIR)
        cd $(CRENGINE_BUILD_DIR) && \
                CC="$(CC)" CXX="$(CXX)" RC="$(RC)" \
```

Maybe something could be done with a pattern rule…

**Comment by poire-z** [comment]
Using `$(THIRDPARTY_DIR)/kpvcrlib/crengine/crengine/src` does not trigger any rebuild for me if I update a file in `src/` (and we should not forget `include/`).

**Comment by poire-z** [comment]
Yep, this one works :) (dunno if we really need to force a recompile if .pattern or .pdb have changed, they are not compiled into libcrengine.so, but loaded on the fly when needed, and reloaded when switching documents.

**Comment by Frenzie** [comment]
Indeed, it's just a double symlink. One from koreader-target/koreader/data to base/build/target/data and one from there to thirdparty/kpvcrlib/etc. Also I just noticed there's already a `CRENGINE_SRC_DIR` defined in Makefile.defs:

```
CRENGINE_SRC_DIR=$(CURDIR)/$(THIRDPARTY_DIR)/kpvcrlib/crengine
```

PR coming soon.

**Comment by cramoisi** [comment]
(ok, no surprise I was lost :))

**Comment by poire-z** [comment]
Should I make a PR with MAX_PATTERN_SIZE 9 > 16 + FORMATTING_VERSION_ID++ ? (dunno if and how I can hijack #145)
Quickly tested with French, and looks like a low risk: it will allow us to check on our devices over the weekend?

**Comment by Frenzie** [comment]
You can clone/pull/whatever from @cramoisi's branch (provided he enabled write access for us) and then push there.

**Comment by poire-z** [comment]
Thanks, it worked.
Feels strange impersonating @cramoisi :)

**Comment by cramoisi** [comment]
@poire-z @Frenzie ; I can test some of the pattern I did yesterday (en-US and en-GB will be easy) but I don't have books in bu, cz and fi ...

**Comment by poire-z** [comment]
I guess you should mainly test your new French.pattern. Assuming you did the other languages files the same way you did the French one, they should be fine. And other people will validate these when you commit them in next nightlys.
Unless we let your files sit here for a while so interested people can grab them and test them.
@jchtt : can you give us your feedback about hyph_de_DE.zip from https://github.com/koreader/koreader/issues/3848#issuecomment-380801455 with today's nightly (>= 1658) ?

**Comment by cramoisi** [comment]
Yes, I did the other the same way. In fact, French was the more complicated as it had some ' and - when other languages doesn't have them.

I must add, regarding the OP, these pattern should permit people who want hyphenate some word in a precise place to do it, (with the example in the OP) adding `<pattern>extra5ordinary</pattern>` at the end of the .pattern file (before the closing tag) should do the trick

**Comment by poire-z** [comment]
Is it ok if I don't see any change (in nb of total pages and where hyphens are placed comparing a few pages) when I switch between my old French.pattern and yours from https://github.com/koreader/koreader/issues/3848#issuecomment-380829671 ? You wrote there _You shouldn't see any differences with koreader's file (all this for this :))_, but there's still more lines in yours.

**Comment by cramoisi** [comment]
yes. My pattern takes into account some utf-8 chars and some rules that are more rarely used or only for some words. You could see them in the [diff file attached there](https://github.com/koreader/koreader/issues/3848#issuecomment-380780674)

**Comment by poire-z** [comment]
Don't worry about them, @Frenzie added these recently to debug the AppImage stuff. So we get them on Kobo when it tries to detect the platform (by checking load of libraries), so we get these on our Kobo because we don't have any SDL or libblitbuffer.so.
Is you menu id plus log still there then ?

**Comment by cramoisi** [comment]
yes , it's still there ...

**Comment by cramoisi** [comment]
If you want to see the diffrence, the only way would be to made a epub with words concerned by the patterns difference or, perhaps to open a very big book (i will try with À la recherche...)

**Comment by poire-z** [comment]
It's fine if I don't see them :) it means it works as well as before, so no surprise.

**Comment by cramoisi** [comment]
and with the new value, the text seems to me better to the eyes. No more "large" spaces which were the results of ignored patterns

**Comment by cramoisi** [comment]
@poire-z you should add `<pattern> jus5qu </pattern>` at the end of your file.

**Comment by poire-z** [comment]
@cramoisi : aren't you the Master Of The French.pattern now ? :)
I thought you were going to replace it with yours (_My pattern takes into account some utf-8 chars and some rules that are more rarely used or only for some words_) ?

Or you ask me that just for testing? (and so, what to test/expect?)

**Comment by cramoisi** [comment]
@poire-z I will replace it but I haven't got enough time this week-end to test it as I wanted, but i'll PR it tomorrow if all goes well. So, yep, you can test it now if you want.
I want to ask if you think the spacing between words is alright as it is. I can't do much about that anyway...

But for my personal use I overrode a lot of rules that are not "correct when you are from university" (as were people which makes the tex pattern are from) but made a nicer display as far as I'm concerned.

The fact is the tex rules are a first process in the document. Ideally (if you have to layout a book) you have to rescan the book multiple times for ruling out all the name, titles, and things like that. Idem from words from another langage, where the langage's rules have to be applied. At the end, the nice layout is not compute, it's people work. Koreader won't be able to do it, so it's wrong in a lot of point anyway, even with a good set of rules and since I hate large spaces between words; I made it for myself. I attach it there for people who want to try (but they won't be add in the PR).

[finales muettes.txt](https://github.com/koreader/koreader/files/1915482/finales.muettes.txt)

**Comment by jchtt** [comment]
@poire-z @cramoisi I tested the new German pattern. Sorry it took so long, I wanted to do it on my regular computer so I could resize the window, but building was a bit of a headache.

Anyway, it works great, although I don’t see many differences compared to the old pdb file. I guess I could go in and edit out any short splits, but options for the minimum number of characters to split would still be nicer.

**Comment by cramoisi** [comment]
@jchtt There was no differences to see :) But the fact is the pattern files are text file which can be easily read and edited, so I guess it's better than old pdb files.

> go in and edit out any short splits

It's kinda the opposite, you have to override the pattern for long words by saying where to hyphen them with high odd and even value (go for 7 and 8), something like : 

`ex8tra7or8di8na8ry,`

But it will be a lot of work for not a very good result. It would be best to have a min left and right value passed to hypman via the pattern file. @poire-z what about the possibility to ignore a pattern if the cut is made before a certain length on the left and right of a word ? Let's say I made left / right at length 3, for a word like ' invention ', the result would be to ignore the match 've 100' : 

TRACE word to hyphenate: ' invention '
DEBUG Pattern matched:  i 004 on  invention  000000000000
DEBUG Pattern matched: ve 100 on vention  000000000
DEBUG Pattern matched: ti 100 on tion  000000
TRACE Hyphenate: in-ven-|tion  i4n1-v0e0n1-|t0i0o0n0

**Comment by poire-z** [comment]
I've been reading over the weekend with yours from from friday, and I wasn't suprised in anything - so that's my validation :) (again, I'm not as sensitive as you about hyphenation, so you won't get quality review from me on that :| )
I've never really been bothered by the spacing between words. I sometimes notice such lines, which makes me stop to wonder why - but nothing really bothering. Didn't happen over the weekend I thing, so may be it's because of you (the number of pages of the book I was reading haven't changed thus).

Btw, you could add in languages.json the current conservative one as the default for French, and add your extreme one as an alternative non-default one. Dunno if it's a good idea to do that thus (@Frenzie thoughts ?) - in alphabetical order, or tweaked ones at the end of the list? Or just let them on the wiki for interested people to download?

**Comment by poire-z** [comment]
About min length on each side, I think you could get to have that by tweaking this line:
https://github.com/koreader/crengine/blob/b3a261270596872ecb2c7deeaa9627c0cc84f286/crengine/src/hyphman.cpp#L792 `for ( p=len-3; p>=1; p-- ) {`

With `for ( p=len-4; p>=2; p-- ) {` kar-mique was kept, but mi-neur was avoided.
With `for ( p=len-5; p>=3; p-- ) {` both were avoided.

Just to see if it's there these LEFTHYPHENMIN and RIGHTHYPHENMIN values should be plugged.

**Comment by cramoisi** [comment]
> Just to see if it's there these LEFTHYPHENMIN and RIGHTHYPHENMIN values should be plugged.

Great ! It's kinda the only feat which miss in koreader hyphenation at this point.

If the german pattern is ok (as @jchtt is saying), we could get rid of the pdb and put it in place. Only thing is I'm not sure to be able to do that, I need a lot of training with github ...

**Comment by poire-z** [comment]
> Just to see if it's there these LEFTHYPHENMIN and RIGHTHYPHENMIN values should be plugged.

@cramoisi and @jchtt ; can you test and validate this by modifying this line in hyphman.cpp to your liking (you both have the emulator I think)?

In case it works and it's just a matter of adding these, where should these settings come from?
Should they be user preferences (per language or global and the same would apply to any language?), or somehow hardcoded in the pattern file for a language (like patterns are)?

In my old source for French.pattern from LibreOffice, I read in `hyph-fr.tex`:
```
% Before a new "pattern-revolution" starts,
% please try to follow some guidelines if possible:
%
% - \lccode is *forbidden*, and I really mean it
% - all the patterns should be in UTF-8
% - the only "allowed" TeX commands in this file are: \patterns, \hyphenation,
%   and if you really cannot do without, also \input and \message
% - in particular, please no \catcode or \lccode changes,
%   they belong to loadhyph-foo.tex,
%   and no \lefthyphenmin and \righthyphenmin,         <------
%   they have no influence here and belong elsewhere   <------
```
so I guess these should not be in our .pattern files either (which I would hate to have to parse in C :)

**Comment by cramoisi** [comment]
@poire-z as it is in the setting of .tex and openoffice .dic, by langage file. Tex said it's not belong there because texts processor has to take it into account and there is an epub3 CSS rule about this. But it's seems to me more interesting to have that in the pattern file which is directly related to it , and per langage than in each books...

> for ( p=len-4; p>=2; p-- )

so the first value is left, and the second is right ?

**Comment by cramoisi** [comment]
@poire-z 
> for ( p=len-3-RIGHTHYPHENMIN ; p>=1+LEFTHYPHENMIN ; p-- ) {

you mean p=len-(3-righthyphenmin) or p=len-(3+righthyphenmin) ? :)

**Comment by poire-z** [comment]
I mean as I wrote - or as you wrote it with parens: `p=len-(3+righthyphenmin)`

**Comment by cramoisi** [comment]
@poire-z : it's working as intended !

**Comment by cramoisi** [comment]
gosh, i change the total stupid me...
but it works the same with adding explicitly the value

**Comment by poire-z** [comment]
To see the kind of logs I pasted last week?:
Change https://github.com/koreader/crengine/blob/b3a261270596872ecb2c7deeaa9627c0cc84f286/crengine/src/hyphman.cpp#L17 to `if 1`
and that one https://github.com/koreader/koreader-base/blob/7a294b4b2625cea4bc2618f00b40256f5c14e4de/cre.cpp#L21 to `#define DEBUG_CRENGINE 1` too.

**Comment by cramoisi** [comment]
thanks ! I got the first (easy enough) was searching for the second.
What do you mean exactly (do as I'm six) with your +1 - 1 thing ?

**Comment by poire-z** [comment]
I don't really know what the default value looks like it is with `for ( p=len-3; p>=1; p-- ) {` 
`1` or `2` ?
I'm not really sure.
The code looks like 1. So I would need to substract 1 to the provided values.
But I only get hyphenation with minimum 2 chars on on side when browsing a few page. So if it is 2, I would need to substract 2 to the provided values.

**Comment by cramoisi** [comment]
2018/04/16 19:44:27.6661 TRACE word to hyphenate: ' particulier '
2018/04/16 19:44:27.6662 DEBUG Pattern matched: pa 100 on particulier  0000000000000
2018/04/16 19:44:27.6663 DEBUG Pattern matched: ti 100 on ticulier  0000000000
2018/04/16 19:44:27.6663 DEBUG Pattern matched: cu 100 on culier  00000000
2018/04/16 19:44:27.6663 DEBUG Pattern matched: li 100 on lier  000000
**2018/04/16 19:44:27.6663 TRACE Hyphenate: par-ti-c|u-lier  p0a0r1-t0i1-c0|u1-l0i0e0r0**

the log seems off, so it somewhat needs adjustement...

**Comment by cramoisi** [comment]
it's not because the min length size of a word for hyphenation to be processed is 2 ?

**Comment by poire-z** [comment]
> the log seems off, so it somewhat needs adjustement...

I see, that quick trick was bad. We shouldn't skip these first or last chars in the loop: we just need to put 0 in the returned flags for them.

**Comment by cramoisi** [comment]
I said that because of the position of the pipe but i'm not really sure.

2018/04/16 19:34:43.3867 TRACE word to hyphenate: ' accourions '
2018/04/16 19:34:43.3867 DEBUG Pattern matched:  a 004 on  accourions  0000000000000
2018/04/16 19:34:43.3867 DEBUG Pattern matched: co 100 on courions  0000000000
2018/04/16 19:34:43.3867 DEBUG Pattern matched: ri 100 on rions  0000000
2018/04/16 19:34:43.3867 TRACE Hyphenate: ac-cou-rio|ns  a4c1-c0o0u1-r0i0o0|n0s0
2018/04/16 19:34:43.3867 TRACE word to hyphenate: ' traversait '
2018/04/16 19:34:43.3868 DEBUG Pattern matched: tr 120 on traversait  000000000000
2018/04/16 19:34:43.3868 DEBUG Pattern matched: ra 100 on raversait  20000000000
2018/04/16 19:34:43.3868 DEBUG Pattern matched: ve 100 on versait  000000000
2018/04/16 19:34:43.3868 DEBUG Pattern matched: sa 100 on sait  000000
2018/04/16 19:34:43.3868 TRACE Hyphenate: tra-ver-sa|it  t2r0a1-v0e0r1-s0a0|i0t0

**Comment by poire-z** [comment]
Mhh, these logs are logged before we reach our `for ( p=len-3; p>=1; p-- ) {` ... and they look indeed different than before, and wrong...
The actually differ from what I see on the screen, which looks correct.

**Comment by poire-z** [comment]
Looking at the code, he minus sign means "hyphen allowed here".
And the bar is put where some maxWidth is reached, which is were the caller will go to find a cut.
So, it looks ok, isn't it?

**Comment by cramoisi** [comment]
@poire-z it's ok !

2018/04/16 20:28:29.9798 TRACE word to hyphenate: ' supercalifragi '
2018/04/16 20:28:29.9798 DEBUG Pattern matched:  supercalifragi  00555555555555500 on  supercalifragi  00000000000000000
2018/04/16 20:28:29.9799 TRACE Hyphenate: s-u-p-e-r-c-a-l-i-f-r-a-g-i  s5-u5-p5-e5-r5-c5-a5-l5-i5-f5-r5-a5-g5-i0

for ( p=len-3; p>=1; p-- )
2018/04/16 20:29:30.2410 TRACE word to hyphenate: ' supercalifragi '
2018/04/16 20:29:30.2410 DEBUG Pattern matched:  supercalifragi  00555555555555500 on  supercalifragi  00000000000000000
2018/04/16 20:29:30.2410 TRACE Hyphenate: s-u-p-e-r-c-a-l-i-f-r-a-g-|i  s5-u5-p5-e5-r5-c5-a5-l5-i5-f5-r5-a5-g5-|i0

=> the left is at 2 min
=> the right is at 2 min


for ( p=len-3-1; p>=1+1; p-- ) {


=> the left is at 3 min
=> the right is at 3 min

So I guess the first value is always 1 less and the second always 1 more, but it's working

**Comment by cramoisi** [comment]
( p=len-2; p>=0; p-- ) changes nothing the min right is 2 and the min left is 2 ; which is ok I guess

**Comment by poire-z** [comment]
Indeed. I was just tested that too :)

**Comment by poire-z** [comment]
OK, I'll try tomorrow to find a way to push these settings from the lua side up to down there.

**Comment by jchtt** [comment]
@poire-z I also checked, it’s looking good! Testing through different values gives exactly the right behavior for
```
for ( p=len-3-(RIGHTHYPHENMIN-2) ; p>=1+(LEFTHYPHENMIN-2) ; p-- ) {
```
provided both are at least 2 (and of course this can be simplified :)).

Also, maybe I’m just weird by preferring longer breaks, I wouldn’t generalize to all German readers here. But it’s nice to have that as a tuning option. As to where it should be, I’m usually all for having it easily accessible and not too hidden away, so one can easily experiment with it and change it, if it’s not too much work. Maybe one could carve off a little menu space in the Hyphenation menu?

Really awesome work for finding it so quickly! :+1:

**Comment by poire-z** [comment]
OK, implemented by https://github.com/koreader/crengine/pull/151 and #3890. Seems to work as expected.
If you want to test them on the emulator:
```
cd /tmp/
wget https://github.com/koreader/crengine/pull/151.diff
wget https://github.com/koreader/koreader/pull/3890.diff
cd .../koreader/
cd .../koreader/base/thirdparty/kpvcrlib/crengine
patch -p1 < /tmp/151.diff
cd .../koreader/
patch -p1 < /tmp/3890.diff
./kodev build
```
When done playing, and before your next git pull, revert them:
```
cd .../koreader/base/thirdparty/kpvcrlib/crengine
git reset --hard
cd .../koreader/
git reset --hard

```

**Comment by Frenzie** [comment]
I generally find it simpler to do something like

```bash
# first commit changes to branch, stash or discard, not pictured

# reset to common starting point, might also include something like HEAD~100
git checkout master

# create new test branch, normally with a more descriptive name
git checkout -b test-branch

# pull in the changes
git pull https://github.com/koreader/crengine.git

# although in practice more like `git pull https://github.com/poire-z/crengine.git branch-name`

# etc.
./kodev build
```

Then you can just git checkout master/your own branch afterward plus you can still switch back & forth quickly.

Even when I do manual patching I still do it in a Git branch for the same convenience. Git rocks. :-P

**Comment by NiLuJe** [comment]
I also like git's stash support, but that may be because of previous forced exposure to quilt ;p.

**Comment by jchtt** [comment]
@poire-z I really like the menu! Unfortunately, it only seems to be working when “Algorithm” is selected as the hyphenation dictionary, not when I select either the English or the German hyphenation dictionary.

**Comment by poire-z** [comment]
Mhhh, it works for me, with either French, German (new .pattern), English or Italian (old .pdb files).

What do you mean with _not working_ ?
Do you see the underlying text being re-rendered somehow (= moving :) or not at all? In that case, is there some hyphenation applied ? With the default value of 2?
Or are there just the length limits badly applied and you have words under the limits you set?

**Comment by jchtt** [comment]
In the debug output, I see some re-calculating going on, but the text stays the same. That means that there are words that get broken that are under the limit. It just stays at the default of 2.

**Comment by cramoisi** [comment]
sorry, I was sleeping. Working for me too.

**Comment by cramoisi** [comment]
@poire-z no possibility to got a right-hyphen to 1 ? (ok, I'm obsessed)

**Comment by poire-z** [comment]
Thanks for the tests!

> no possibility to got a right-hyphen to 1 ? (99,9% of the rules sets the thing to 2 when applied.)

What? Can you re-explain like if I were a 6 years old :)

**Comment by cramoisi** [comment]
@poire-z since crengine doesn't take into account words that have a quote inside them and need special hyphenation rules. Because of that, I had to make special rules for taking them into account since the actual used by ooo, tex ,etc are ignored. The fact is (at least) two of them demand a hyphen at 1 char of the end of the special (truncated) word, so these won't be hyphenate at all with a limit of right 2.

When I said putting the right (or left) limit at 1 shouldn't be a problem, it is because all the rules in the .pattern prevent hyphenation at <2 left and right except when it is explicitly said or override. It's the same for all the languages as the general rules are the same (syllabes, morphologie, composition, etc.)

**Comment by poire-z** [comment]
You could try by changing into hyphman.cpp, for the line we talked about previously:
`for ( p=len-3; p>=1; p-- ) ` to `for ( p=len-2; p>=1; p-- ) `
and removing the checks for our new settings I added yesterday.
```c
--- a/crengine/src/hyphman.cpp
+++ b/crengine/src/hyphman.cpp
@@ -811,11 +811,7 @@ bool TexHyph::hyphenate( const lChar16 * str, int len, lUInt16 * widths, lUInt8

     bool res = false;
     int p=0;
-    for ( p=len-3; p>=1; p-- ) {
-        if (p < HyphMan::_LeftHyphenMin - 1)
-            continue;
-        if (p > len - HyphMan::_RightHyphenMin - 1)
-            continue;
+    for ( p=len-2; p>=1; p-- ) {
         // hyphenate
         //00010030100
         int nw = widths[p]+hyphCharWidth;
```
With this, with the Italian hyph dict I had set, I got:
 ```
procède en éliminan-
t toute interférence
```
May be it's just these `len-3` and `p>=1` that artificially to 2 (well, dunno, if there are other effects, but it does not crash).
Dunno if there would be more bad effect if we set this limit from 2 to 1...

**Comment by poire-z** [comment]
Other question: is there some academic/experimental prefered hyphenation choice for that kind of word:
```
vo-
lonté
                   or
volon-
té
```
Do the rules give a prefered one?
(Just thinking if I were to change the left/right hyphen min values, what would be best: left=2/right=3 or left=3/right=2 ? or I just shouldn't bothered and trust the rules?)

**Comment by cramoisi** [comment]
> Dunno if there would be more bad effect if we set this limit from 2 to 1

pretty sure it won't be bad for the pattern i've checked but I asked more because it bothers me to have all the rules ok and running **except** two of them (and each time for long words)

**Comment by poire-z** [comment]
Looks like we could support 1 as the min value (at least, it hasn't crashed on me yet :)

Is that what you'd want:
<kbd>![image](https://user-images.githubusercontent.com/24273478/38931548-d9e7987e-4313-11e8-88b0-c96b756479a0.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/38931579-f4f40d78-4313-11e8-9011-82192ff6f587.png)</kbd>

**Comment by cramoisi** [comment]
\o/

2018-04-18 14:26 GMT+02:00 poire-z <notifications@github.com>:

**Comment by poire-z** [comment]
Ok, I'll get some PRs updated for you to test this evening.

Just in case you got familiar with reading these (I'm not) (although I think they should be the same as before, the new stuff happens after :)
```
2018/04/18 15:45:16.6843 TRACE word to hyphenate: ' aujourd '
2018/04/18 15:45:16.6843 DEBUG Pattern matched:  aujourd  0005000500 on  aujourd  0000000000
2018/04/18 15:45:16.6843 DEBUG Pattern matched:  aujourd 000500050 on  aujourd  0005000500
2018/04/18 15:45:16.6843 DEBUG Pattern matched:  a 004 on  aujourd  0005000500
2018/04/18 15:45:16.6844 DEBUG Pattern matched: j 10 on jourd  5000500
2018/04/18 15:45:16.6844 TRACE Hyphenate: au-jour-d  a4u5-j0o0u0r5-d0
```
Guess we can stay with the default 2/2 for some time, and you and I will be able to set 2/1 via the widget to test for some time. Then we can put the french defaults it in language.json if they are good.

**Comment by cramoisi** [comment]
> they should be the same as before 
They are !
> Guess we can stay with the default 2/2 for some time, and you and I will be able to set 2/1 via the widget to test for some time.

no problem. so you'll pull the code for 1 in the hyphman but not in the widget and so we'll just have to modify the lua file ? (if not, I will need help to compile my own kobo binary, some libraries miss on my macbook)

**Comment by poire-z** [comment]
I'll push a followup PR for hyphman (I merged the previous one early this morning).
I'll change the min in the lua widget from 2 to 1, and update this PR (not yet merged) so anyone can test (when it's merged) and see (if it doesn't crash for us, it shouldn't elsewhere). The defaults and the "Use language defaults" will stay at 2.
Have a quick test this evening on the emulator, and if it looks fine, we'll merge and bump so it's in tomorrow or next day nightly - don't bother with compiling for arm...

**Comment by poire-z** [comment]
I guess we should also do something about the fact that when starting koreader and initializing crengine (it starts with a dummy HTML doc before it reads our provided book), 2 hyph dictionary are loaded and parsed:
```
2018/04/18 16:50:56.9231 INFO HyphDictionaryList::open(./data/hyph/)
2018/04/18 16:50:56.9239 INFO 31 items found in hyph directory
2018/04/18 16:50:56.9239 TRACE HyphDictionaryList::activate(Russian_EnUS_hyphen_(Alan).pdb)
2018/04/18 16:50:56.9240 INFO Selecting hyphenation dictionary ./data/hyph/Russian_EnUS_hyphen_(Alan).pdb
2018/04/18 16:50:56.9782 DEBUG Pattern: 'hyph' - 00300
2018/04/18 16:50:57.1277 TRACE word to hyphenate: ' please '
2018/04/18 16:50:57.1279 TRACE Hyphenate: pleas|e  p2l2e0a2s2|e0
2018/04/18 16:50:57.1282 DEBUG PROP_HYPHENATION_DICT = English_US_hyphen_(Alan).pdb
2018/04/18 16:50:57.1282 DEBUG Changing hyphenation to English_US_hyphen_(Alan).pdb
2018/04/18 16:50:57.1282 TRACE HyphDictionaryList::activate(English_US_hyphen_(Alan).pdb)
2018/04/18 16:50:57.1307 INFO Selecting hyphenation dictionary ./data/hyph/English_US_hyphen_(Alan).pdb
2018/04/18 16:50:57.1825 DEBUG Pattern: 'hyph' - 00300
(and there comes the one I selected previously)
04/18/18-16:50:57 DEBUG CreDocument: set hyphenation dictionary @algorithm
2018/04/18 16:50:57.5211 DEBUG PROP_HYPHENATION_DICT = @algorithm
2018/04/18 16:50:57.5211 DEBUG Changing hyphenation to @algorithm
2018/04/18 16:50:57.5211 TRACE HyphDictionaryList::activate(@algorithm)
2018/04/18 16:50:57.5211 INFO Turn on algorythmic hyphenation
04/18/18-16:50:57 DEBUG CreDocument: set hyphenation left hyphen min 1
04/18/18-16:50:57 DEBUG CreDocument: set hyphenation right hyphen min 1
```
may be by adding a `Dummy.pattern` (not referenced in language.json) with zero pattern, and using that one in crengine defaults and in cre.cpp defaults?

**Comment by cramoisi** [comment]
thanks ! @none seems right because at this moment nothing has to be hyphen. The selection of English_US_hyphen_(Alan).pdb is not related to the cr3.ini ? there is a key=value with it.

**Comment by poire-z** [comment]
Yes, it's in cr3.ini, which I'm not sure is used.
A hyph dict is set:
(1) once by hyphman.cpp when initializing the directory, a quick check for the directory validity, that it contains its default dictionary.
(2) then, later, in cre.cpp: we set some defaults.
(3) then, later, when loading document: the doc (.sdr) defined dict, or the user defined dict, or one base on document language.

(2) is `English_US_hyphen_(Alan).pdb`
(1) was `Russian_EnUS_hyphen_(Alan).pdb`. I just set it too to `English_US_hyphen_(Alan).pdb`, so it's not reloaded when 2) happens.
`English_US_hyphen_(Alan).pdb` is not small, but well, all that happens only once on first credocument opening in a session.

**Comment by cramoisi** [comment]
@Frenzie I said that because I can't cross-compiling on my MacBook (not that it can't ; but i can't figure out a solution to have  gcc-arm-linux-gnueabihf  / g++-arm-linux-gnueabihf )

**Comment by Frenzie** [comment]
I assume Macs run Docker, which might be the easiest solution. (Well, except Docker itself is arguably not that easy.)

@onde2rock wrote a little something about how you could use it here though it's not quite right for your use case: https://github.com/koreader/koreader-base/pull/640#issuecomment-379759220 (the `-v` is volume mount)

You can grab the official Kobo nightly build image with `docker pull frenzie/kokobo:0.0.4` but note that I'll be creating a 0.0.5 today for @NiLuJe 's sdcv fix.

All that being said, I'm not sure if I'd recommend getting to know Docker for more casual use. But it might not be any harder than figuring out how to set up a cross-compile toolchain on Mac OS and ultimately more versatile. You shouldn't need to figure out too much because you only need to pull and use the containers that I create.

It's a matter of perspective. To me Docker is easier to set up and use than Vagrant, not to mention faster. It's easier to set up than a chroot. And finally, it's also a lot easier to use than systemd-nspawn. Basically there's probably nothing better *and* easier than Docker right now. But it's still a pretty complex kind of easy.
